### ISSUE TITLE
some comments on block/sector size ( ch 14 )

### ISSUE THREAD

**Issue by gf-mse** [shichao-an/notes#7]
to https://github.com/shichao-an/notes/blob/master/docs/lkd/ch14.md  :
_I'm not sure whether that fits as book comments, but since it was probably the best introduction in the matter that I could quickly find online, I want to share my two cents, that one may or may not find useful:_

----

Suppose we need to find out what's the optimal buffer size for reading from a device. Or have special needs -- e.g. want to tune a ZFS filesystem recordsize for a custom on-disk database -- may be even with enabled deduplication, what will make our needs really special.

So we wonder how the system reads and writes data from/to the device. These i/o _buffers_ would be directly related to sector or block sizes.

----

As for device block sizes -- they, of course, first come from the device itself ; look at 

    /sys/block/sda/queue :

        hw_sector_size
        logical_block_size
        physical_block_size

That would be the same as `smartctl` or `hdparm` would tell us:

    smartctl -a /dev/sda    # grep -i block
    hdparm -I /dev/sda     # grep -i block

"Physical" and "logical" here would differ e.g. for [Advanced Format](https://wiki.archlinux.org/index.php/Advanced_Format) disks ( and we'll be more interested in `physical_block_size`, if it's different ).

----

Next, there's something reported by `man 2 stat` ( `struct stat -> st_blksize` ) : 

    stat /etc/passwd ... -> ... IO Block: 4096

This essentially comes from the filesystem. ( In short,  [fs/stat.c](https://github.com/torvalds/linux/blob/master/fs/stat.c) shows that `st_blksize` comes from the inode, the inode [gets it from the superblock](https://github.com/torvalds/linux/blob/master/fs/inode.c),  and the superblock [normally has it as a power of two capped at PAGE_SIZE](https://github.com/torvalds/linux/blob/master/fs/block_dev.c), i.e. the memory page size. )

So basically this boils down to a memory page size. Good to know. 

----
PS. This is for example how `cat` determines the optimal buffer size : 

 * https://unix.stackexchange.com/questions/17295/when-is-dd-suitable-for-copying-data-or-when-are-read-and-write-partial
 * https://unix.stackexchange.com/questions/245499/how-does-cat-know-the-optimum-block-size-to-use

-- and it does not perform that bad at all:
 * https://unix.stackexchange.com/questions/9432/is-there-a-way-to-determine-the-optimal-value-for-the-bs-parameter-to-dd/9492#9492
